File: src/main/java/lantern/WordPath.java
package lantern;

import org.eclipse.collections.api.list.ImmutableList;

public record WordPath(ImmutableList<Pt> steps, String word) {
    boolean visited(Pt pt) {
        return steps.contains(pt);
    }

    WordPath addStep(Pt pt, char ch) {
        return new WordPath(steps.newWith(pt), word + ch);
    }
}

File: src/main/kotlin/lantern/WordPath.kt
package lantern

data class WordPath(val steps: List<Pt>, val word: String) {
    fun visited(pt: Pt): Boolean = steps.contains(pt)

    fun addStep(pt: Pt, ch: Char): WordPath {
        return WordPath(steps.plus(pt), word + ch)
    }
}
